sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,379 @@
+/// containing kernels from that set.
+/// Access must be guarded by the \ref Sync::getGlobalLock()
+std::unordered_map<SymbolSetId,
+                   std::unique_ptr<std::vector<RTDeviceBinaryImageUPtr>>>
+    m_DeviceImages;
+
+using StrToKSIdMap = std::unordered_map<string_class, SymbolSetId>;
+/// Maps names of kernels from a specific OS module (.exe .dll) to their set
+/// id (the sets are disjoint).
+std::unordered_map<OSModuleHandle, SymbolSetId> m_SymbolSets;
+```
+Assume each device image represents some combination of symbols and different
+device images may contain only exactly the same or not overlapping combination
+of symbols. If it is not so, there can be two cases:
+  - Symbols are the same. In this case it doesn't matter which device image is
+  taken to use duplicated symbol
+  - Symbols are not the same. In this case ODR violation takes place, such
+  situation leads to undefined behaviour. For more details refer to
+  [ODR violations](#ODR-violations) section.
+
+Each combination of symbols is assigned with an Id number - symbol set Id.
+A combination of symbols can exist in different formats (i.e. SPIR-V/AOT
+compiled binary and etc).
+`m_DeviceImages` maps an Id number to an array with device images which represent
+the same combination of symbols in different formats.
+`m_SymbolSets` contains mapping from symbol name to symbol set Id for each OS
+module (.exe/.so/.dll).
+`std::unordered_map` allows to search and access its elements with constant-time
+complexity.
+
+Before compilation of device image to execute a kernel RT checks if the image
+contains any import information in its properies and if it does, then RT
+performs device images collection in order to resolve dependencies.
+
+Ids of all needed symbol sets are found. This is done by iterating through
+`m_SymbolSets` map, i.e. iterating through all available OS modules without
+predefined order and searching for first unresolved symbol in list of imports
+set of target device image. Once device image that contains first symbol is
+met, remaining exported symbols are checked in found image and if
+they match some imported symbols then these matched symbols will be marked as
+resolved. The procedure repeats until all imported symbols are resolved.
+For each found symbol set Id program cache is checked in case if
+necessary set of `SYCL_EXTERNAL` functions has been compiled and if it is true,
+then compiled device image will be re-used for linking.
+Otherwise device image containing required symbols set will be compiled and
+stored in cache.
+
+#### Program caching
+
+Existing support for device code caching is re-used to cache programs created
+from device images with SYCL external functions and linked device images with
+imports information.
+
+##### In-memory cache
+
+Programs that contain only `SYCL_EXTERNAL` functions will be cached only in
+compiled state, so they can be linked with other programs during dependency
+resolution.
```

Comment:
```suggestion
Programs that contain only `SYCL_EXTERNAL` functions will be cached only when exported 
functions are used by images with kernels and identified by `main` image Id.
So once two kernel from different images use functions from the same `library` image it will
be duplicated in cache after linking with main programs during dependency
resolution.
```

Reply:
Okay, there can be two cases when `SYCL_EXTERNAL`-only images are cached in in-memory cache:
- Inside the `main` image, i.e. after it is linked with all its dependencies (the case that you described in your suggestion)
- Non-linked `SYCL_EXTERNAL`-only images in compiled state. It can be done when some "main" image requested some `SYCL_EXTERNAL`-only image as dependency, we created a program for `SYCL_EXTERNAL`-only image, compiled it and stored in cache, so we don't have to do it again when some other `main` requests the same `SYCL_EXTERNAL` image. (the case that I tried to describe here). The idea was taken from existing cache for standard device libraries (see https://github.com/intel/llvm/blob/86716c542268b8a1ad104c85b783cfc6cb593cc9/sycl/source/detail/context_impl.hpp#L177).

Does it make sense?